### PR TITLE
Charge bionics from UPS via Cable Charger System CBM + bugfixes for solar backpacks

### PR DESCRIPTION
**Pull request by nexusmrsep**
#### Summary
```SUMMARY: Features "Charge bionics from UPS via Cable Charger System CBM"```
Also some bugfixes on solar backpacks that had to be done along the way.

#### Purpose of change
There is a lot of discussion lately about the validity of certain methods of power generation for the CBM by CBMs. This led to some bionics being obsoleted, as there is a limit to what you can put inside a body that can generate power, because those things are usually big and/or dangerous.

In all this I noticed that one obvious path was overlooked, or alt least not discussed extensively.

While it is natural that many power-related CBMs are focused on mobility, but the prototypes and most early versions must have used Cable Charger System CBM extensively. With development, the high-end versions might have gone off-line, while civilian versions could as well use Cable Charger System to pull the energy from the grid. You'd plug yourself in for the night, like you plug in your smartphone or an electric car.

There is a middle ground between using grid (now: vehicles) and full mobility of integrated CBMs. This middle ground are external power generating devices or power storage devices that can by plugged in by cable, and carried in inventory.

First of a kind, that I also had a privilege to introduce, were solar backpacks. Wear one, unfold, plug in and get some power with time.

So I though: if Universal Power Supply was designed to power mobile weapon platforms, then why would it not power bionic systems? Power is power, and a soldier / bionic operator could just carry a plugged UPS to power himself in the field. External power, high capacity, universal standardized connectors, mobility. Can't see a reason why it wouldn't be implemented before fully integrated power-generation CBMs.

#### Describe the solution
I've done few things here:
- fixed some bugs in solar backpacks regarding to cable connection not working properly (prerequisite of this PR),
- while at it, I rebalanced power output of solar backpack to be in line with power output of solar panel (50W in sunlight) and made it dependent on sunlight; take note: this is a nerf from previous values, but previous values were predating solar panels rebalance,
- made UPS and advanced UPS connectable via jumper cables / heavy-duty cables to Cable Charger System CBM installed in character's body,
- game tracks available options of 'attaching' cable actively, switching them on/off when power source (UPS, solar pack) is available, and graying them out to avoid unsupported connections,
- game tracks loss of cable or power source

#### Describe alternatives you've considered
- when using various utility CBMs in the absence of internal power storage, game could allow direct power pull from the UPS
- There are some interactions that need a thought or two, like for example "should a UPS plugged to CCS CBM be available to other systems?". Left at "yes" for now.
- for solar backpacks: there is a place for upgraded solar backpack (one between normal and quantum) for better power output

#### Additional context
If you have a UPS with some juice and Cable Power System CBM you can use jumper cables:
image
image
image
image
image

Or other way around.
It will activate your UPS and you will draw power from it.

image

#### TODO
- [x] ~~main features~~
- [x] some descriptions have to change to hint this new feature
- [x] some more testing need to  be done to check all the paths of attach/drop

### CODE REVIEW

**Comment by Rail-Runner** [comment]
Would it be possible to have an ability to just plug self to a battery directly? Maybe even recharge it using your own power, if you want to stockpile energy that you generate in external bateries.

**Comment by nexusmrsep** [comment]
Battery System CBM is for that I believe.

**Comment by dunstad** [comment]
Since the solar backpack charges you through the Cable Charger CBM, which seems to use jumper cables, would it make sense for the backpack to also allow charging vehicle batteries?

Not that you have to implement that by any means, just curious if the logic checks out or if I've got low electronics skill.

**Comment by nexusmrsep** [comment]
Yes. It's basically an equivalent to one solar panel. To do that in game, some infrastructure would have to be built, but it's plausible.

**Comment by Fris0uman** [comment]
While you're at it could you replace `is_wearing( "solarpack_on" )` with a flag check so that solar cloth can be modded in?

**Comment by nexusmrsep** [comment]
I agree that it would be beneficial, but it would also require JSON entry for epower generation (efficiency) and thus more infrastructure that meets the eye. Simple flag would not be enough. I don't want to bloat this PR with something it's not intended to achieve, so let's maybe put it for later. I also stated that there is a place for some more solar backpack variants and such infrastructure would help achieve that goal. I will place it on my ToDo list, but no promises on "when". I already code less then I wish I would, due to time constraints.

File: src/item.cpp [hunk 1]
```diff
@@ -8203,6 +8205,33 @@ cata::optional<tripoint> item::get_cable_target( player *p, const tripoint &pos
 
 bool item::process_cable( player *p, const tripoint &pos )
 {
+    if( p == nullptr ) {
+        reset_cable( p );
+    }
+    std::string state = get_var( "state" );
+    if( state == "solar_pack_link" ) {
+        if( !p->has_item( *this ) || ( !p->is_wearing( "solarpack_on" ) ||
```
Comment (BevapDin) [inline src/item.cpp]: This may get called with `p` being null.

File: src/item.cpp [hunk 2]
```diff
@@ -8203,6 +8205,33 @@ cata::optional<tripoint> item::get_cable_target( player *p, const tripoint &pos
...
+            return false;
+        }
+    }
+
+    static const item_filter used_ups = [&]( const item & itm ) {
+                    return itm.get_var( "cable" ) == "plugged_in";
+                    };
+
+    if( state == "UPS" ) {
+        if( !p->has_item( *this ) || ( !p->has_item_with( used_ups ) ) ) {
```
Comment (BevapDin) [inline src/item.cpp]:
```suggestion
        if( !p->has_item( *this ) || !p->has_item_with( used_ups ) ) {
```
And needs a check for `p` not being null.

File: src/item.cpp (hunk 1 above)
Comment (nexusmrsep) [reply src/item.cpp]: `if( p == nullptr ) { reset_cable( p ); }` is checked at the very begining of the `process_cable()` 
I'll add `return false;` that should bail out from there.

File: src/item.cpp (hunk 2 above)
Comment (nexusmrsep) [reply src/item.cpp]: As above.

File: src/item.h [hunk 3]
```diff
@@ -2020,7 +2020,8 @@ class item : public visitable<item>
         // Place conditions that should remove fake smoke item in this sub-function
         bool process_fake_smoke( player *carrier, const tripoint &pos );
         bool process_fake_mill( player *carrier, const tripoint &pos );
-        bool process_cable( player *p, const tripoint &pos );
+        bool process_cable( player *carrier, const tripoint &pos );
```
Comment (BevapDin) [inline src/item.h]: The clang-tidy-plugin will revert this change. It wants that parameters have the same name in declaration and in their definition.

Comment (nexusmrsep) [reply src/item.h]: Good to know, it happened to me once and I was baffled what was going on. Too bad the result will be a bit inconsistent `item.h` naming scheme, but as long as it works I don't complain.

Comment (nexusmrsep) [reply src/item.h]: Good to know, it happened to me once and I was baffled what was going on. Too bad the result will be a bit inconsistent item.h naming scheme, but as long as it works I don't complain.

File: src/item.cpp [hunk 4]
```diff
@@ -8247,6 +8277,24 @@ void item::reset_cable( player *p )
...
+bool item::process_UPS( player *p, const tripoint & /*pos*/ )
+{
+    if( p == nullptr ) {
+        erase_var( "cable" );
+        active = false;
+        return false;
+    }
+    bool has_connected_cable = p->has_item_with( []( const item & it ) {
+        return it.active && it.has_flag( "CABLE_SPOOL" ) && ( it.get_var( "state" ) == "UPS_link" ||
+                it.get_var( "state" ) == " UPS " );
```
Comment (BevapDin) [inline src/item.cpp]:
```suggestion
                it.get_var( "state" ) == "UPS" );
```

File: src/item.h (hunk 3 above)
Comment (nexusmrsep) [reply src/item.h]: Good to know, it happened to me once and I was baffled what was going on. Too bad the result will be a bit inconsistent item.h naming scheme, but as long as it works I don't complain.

Comment (nexusmrsep) [reply src/item.h]: I've cross-unified it.

File: src/item.cpp [hunk 5]
```diff
@@ -8201,18 +8203,46 @@ cata::optional<tripoint> item::get_cable_target( player *p, const tripoint &pos
...
 {
-    const cata::optional<tripoint> source = get_cable_target( p, pos );
+    if( carrier == nullptr ) {
+        reset_cable( carrier );
+        return false;
+    }
+    std::string state = get_var( "state" );
+    if( state == "solar_pack_link" || state == "solar_pack" ) {
+        if( !carrier->has_item( *this ) || ( !carrier->is_wearing( "solarpack_on" ) ||
+                                             !carrier->is_wearing( "solarpack_on" ) ) ) {
```
Comment (Night-Pryanik) [inline src/item.cpp]: Unless I'm reading this wrong, the check is `X || ( Y || Y )`. Is this intended?

Comment (nexusmrsep) [reply src/item.cpp]: You're right, I didn't noticed that. It should be quantum backpack in that second case. Will fix when possible.

File: src/bionics.cpp [hunk 6]
```diff
@@ -670,28 +670,49 @@ bool player::activate_bionic( int b, bool eff_only )
...
-        bool has_cable = has_item_with( []( const item & it ) {
-            return it.active && it.has_flag( "CABLE_SPOOL" );
-        } );
-        bool has_connected_cable = has_item_with( []( const item & it ) {
-            return it.active && it.has_flag( "CABLE_SPOOL" ) && it.get_var( "state" ) == "solar_pack_link";
+        std::vector<item *> cables = items_with( []( const item & it ) {
+            return it.has_flag( "CABLE_SPOOL" );
         } );
-
+        bool has_cable = cables.empty();
```
Comment (BevapDin) [inline src/bionics.cpp]:
```suggestion
        bool has_cable = !cables.empty();
```

File: src/item.cpp [hunk 7]
```diff
@@ -8201,18 +8203,46 @@ cata::optional<tripoint> item::get_cable_target( player *p, const tripoint &pos
...
+            return false;
+        }
+    }
+
+    static const item_filter used_ups = [&]( const item & itm ) {
+        return itm.get_var( "cable" ) == "plugged_in";
+    };
+
+    if( state == "UPS" ) {
+        if( !carrier->has_item( *this ) || ( !carrier->has_item_with( used_ups ) ) ) {
```
Comment (BevapDin) [inline src/item.cpp]:
```suggestion
        if( !carrier->has_item( *this ) || !carrier->has_item_with( used_ups ) ) {
```
